---
title: "Untitled"
author: "Fred Boehm"
date: "December 16, 2014"
output: html_document
---

## Revision to original strategy

My earlier idea was to download annotation files from illumina.com. However, I was unsure which files corresponded to our chip for bcgwas. It then occurred to me that some of the originally downloaded files, which Michael got from the Washington University genotyping facility, might have the allele mappings.

Note that we'll also need to see how I coded the alleles when creating the numeric genotypes.


## Downloading annotation files from www.biostat.wisc.edu/~ngroup/bcgwas/SNPB/

```{r}
DATA_DIR <- "../bcgwas-ngroup/data"
```

We downloaded the file "SNP_Data_Gould_Infinium_1M_Omni/SNP_Map.csv"

Now, we read that file into R.

```{r, cache=TRUE, eval=FALSE}
# 'eval=FALSE' means that we don't execute this chunk
smap1<- read.csv(file.path(DATA_DIR,"SNP_Map.csv")) 
head(smap1)
```

It seems that the SNP_Map.csv file contains  insufficient. I'll check the other file, "ForwardStrand_FinalReport.csv"

```{r, cache=TRUE}
smap<- read.csv(file.path(DATA_DIR,"ForwardStrand_FinalReport.csv"), skip=9, colClasses= "character")
names(smap)[1] <- "snp.id"
head(smap)
```

```{r, cache=TRUE, eval=FALSE}
smap2<- read.csv(file.path(DATA_DIR, "test.csv"), skip=9) 
# test.csv is the "head" (first ~50 lines ) of ForwardStrand_FinalReport.csv
head(smap2)
```


Once we have the mapping from A/B to A/C/T/G, we can create a small genotype file from the raw csv file and compare it with the geno.ord.RData file. Note that we need to do this to ensure that we know how we created geno.ord.RData as far as mapping A/B to 0,1,2 genotype classes. It may be that I didn't translate the ACTG to Illumina's AB; rather, I might have arbitrarily chosen an allele as the "A" and then counted for each genotype, the number of "A" alleles.

This raises a bigger question -- can I reproduce the pre-processing steps in the bcgwas analysis?

**I'd like to try to regenerate geno.ord.RData from the source files.**

# Dec 24, 2014

## Reading text file from ngroup/snp directory

```{r, eval=FALSE, cache=TRUE}
read.delim(file.path(DATA_DIR, "snp", "Data")
#install.packages("http://cran.r-project.org/src/contrib/Archive/gdsfmt/gdsfmt_1.0.0.tar.gz", repos=NULL)
```

## Reading in geno.ord.RData 

We are trying to re-create the geno.ord.RData file. We want to compare a portion the current geno.ord.RData file with the file that we'd get from `r head(smap)` object. We also load the snp annotation object that we created in august 2011.

```{r "reading-geno.ord", cache=TRUE, eval=TRUE}
load("../bcgwas-data/EIGENBCGWAS_Data/geno.ord.RData") # loads geno.ord
sann<- pData(getobj("../bcgwas-data/SNP_annotation/illumina.madison.snp.annot.v3.RData"))
```

```{r "split1-defined", cache=TRUE, eval=TRUE}
split1 <- function(x)strsplit(x, "|")
sa <- split1(as.character(smap[1,-1]))
t(sapply(X=sa, FUN=function(x)x[1:2]))
colnames(smap)
newdat <- smap[,c(-11, -23, -32, -33)]
split2 <- function(x) # one row of  data (ie, one snp) is 

```

We then removed all four runs involving N31 & N57 since I don't recall which - N31 or N31.rep and N57.rep or N57 - was used. I'm quite certain that the ".rep" columns were used, but I don't know whether these two are part of the correlation issue.

We then 

```{r, "genoparse-defined"}
split1 <- function(x)strsplit(x, "|")[1:2]
  
split2<- function(x){t(split1(x))}

genoparse <- function(gv.actg) # gv.actg is a character vector with each entry a 2-letter genotype
  {
  foo<- strsplit(gv.actg, "")
  f2<- matrix(ncol = 2, data=unlist(foo), byrow=TRUE)
  o1  <- apply(FUN = function(x)sum(x==f2[1]), X = f2, MARGIN=1)
  other<- f2[f2 != f2[1]][1]
  ref <- f2[1]
  o2 <- c(ref, other)
  out<- list(o1, o2)
  names(out)<- c("genovec", "allele identities")
  return(out)
  }


```

Now, try genocount function.

```{r}
t1 <- c("TT", "TC", "CC", "CC")
genocount(t1)
```

Next steps: 

1. function to compare the genotype vector from geno.ord with that outputted by genocount()
2. after comparing, decide which allele is mapped to 1 and which to zero in geno.ord.



