---
title: "Untitled"
author: "Fred Boehm"
date: "December 16, 2014"
output: html_document
---

## Revision to original strategy

My earlier idea was to download annotation files from illumina.com. However, I was unsure which files corresponded to our chip for bcgwas. It then occurred to me that some of the originally downloaded files, which Michael got from the Washington University genotyping facility, might have the allele mappings.

Note that we'll also need to see how I coded the alleles when creating the numeric genotypes.


## Downloading annotation files from www.biostat.wisc.edu/~ngroup/bcgwas/SNPB/

We downloaded the file "SNP_Data_Gould_Infinium_1M_Omni/SNP_Map.csv"

Now, we read that file into R.

```{r, cache=TRUE}
smap1<- read.csv("data/SNP_Map.csv")
head(smap1)
```

It seems that the SNP_Map.csv file contains  insufficient. I'll check the other file, "ForwardStrand_FinalReport.csv"

```{r, cache=TRUE}
smap<- read.csv("data/ForwardStrand_FinalReport.csv", skip=9)
names(smap)[1] <- "snp.id"
head(smap)
```

```{r, cache=TRUE}
smap2<- read.csv("data/test.csv", skip=9)
head(smap2)
```


Once we have the mapping from A/B to A/C/T/G, we can create a small genotype file from the raw csv file and compare it with the geno.ord.RData file. Note that we need to do this to ensure that we know how we created geno.ord.RData as far as mapping A/B to 0,1,2 genotype classes. It may be that I didn't translate the ACTG to Illumina's AB; rather, I might have arbitrarily chosen an allele as the "A" and then counted for each genotype, the number of "A" alleles.

This raises a bigger question -- can I reproduce the pre-processing steps in the bcgwas analysis?

**I'd like to try to regenerate geno.ord.RData from the source files.**




